### DISCUSSION TITLE
What is the best way to replace kernel when developping?

### DISCUSSION THREAD

**Discussion by rbuchmann77** [OE4T/meta-tegra#997] [Q&A]
Hello,

I'm currently trying some set of kernel configurations and the only way to test it is to build & flash the full image.

I'm using mender. So it's faster than using doflash.sh.
However, yocto takes a lot of time to build the full image (not only the kernel) and the mender bundle.

I'm looking for the faster approach.

The following command is nice to understand the layout:
`lsblk --f --bytes --output NAME,FSTYPE,MOUNTPOINT,PARTLABEL,SIZE,STATE,TYPE`
I guess boot.img is in the partlabel "kernel" or "kernel_b", given the current slot.
Unfortunately, I failed to read/write kernel image from/to those partitions.

**Comment by madisongh** [comment]
There's no single good answer; there are different options, and which one you might choose depends on the specific Jetson model you're using, your other hardware, which L4T version you're using, and your preferred workflow.

One approach is to use an `extlinux.conf` setup, and just copy the kernel and device tree into `/boot/` to test.  This should work with L4T R32.6.x/R32.7.x across all of the Jetsons.  (For Xavier platforms, you have to include `cboot-extlinux` in your image - see the notes [here](https://github.com/OE4T/meta-tegra/wiki/Using-extlinux.conf-on-Xavier-platforms).)  You may have to hand-edit the `extlinux.conf` file with an FDT line to force cboot or U-boot (depending on the target Jetson) to replace the device tree loaded at boot time with the copy associated with the kernel you're testing.

Another approach is to build a BUP payload with your test kernel and device tree and use `tegra-bootloader-update` to install that for testing. The `kernel-bup-payload` (for cboot platforms) and `uboot-bup-payload` (for U-boot platforms) generates a `kernel_only.bup-payload` that you can use for this.  For the TX2 and Xavier, you have to keep track of which boot slot you've booted from and tell `tegra-bootloader-update` to update the kernel/dtb in that slot.

You can also flash just the kernel and DTB to boot partitions via USB, if your builds are not using the `extlinux.conf` kernel loading mechanism.  Exactly what commands you need to do this will vary by Jetson target and by L4T version, but the basic approach is to use the `-c` option to the `doflash.sh` script to have `tegraflash.py`  run specific flashing commands.  If you search around the discussions area here for `doflash.sh`, you'll see some examples.

**Reply by leadercxn** [reply]
Hi @madisongh , I added ` cboot-extlinux ` and ` UBOOT_EXTLINUX_FDT ` in my yocto recipes
<img width="570" alt="image" src="https://user-images.githubusercontent.com/43691182/220023353-887952e8-ed6a-4289-82aa-107cbf16e446.png">

and manually modified `/boot/extlinux/extlinux.conf` and added the dtb file into my system, 
<img width="579" alt="image" src="https://user-images.githubusercontent.com/43691182/220021666-c7afbd88-3d07-4536-acf8-6bd4f0f8c123.png">

but I did not find that the system called the file in the system startup log.
<img width="847" alt="image" src="https://user-images.githubusercontent.com/43691182/220023993-69c8b742-90ff-42b6-b8f0-7e6bab1c6327.png">

HW：Xavier NX
Branch: kirkstone-l4t-r32.7.x

**Reply by leadercxn** [reply]
In fact, I wanted to debug the device tree by replacing the dtb files, rather than burning the entire system, to make the development phase more efficient.

**Reply by leadercxn** [reply]
Oh, I should set these variable  `UBOOT_EXTLINUX_FDTDIR `  and  `UBOOT_EXTLINUX_FDT` in `cboot-extlinux-config.bbclass` file.
